### PR TITLE
Room availability

### PR DESCRIPTION
**Pull request by pl0ss**
https://github.com/neuland-ingolstadt/neuland.app/issues/142

### Summary
🐛🌐🌍

This pull request adds a feature to show the availability information of rooms on the map view of the Rogue THI app. It implements a new function `getRoomAvailability` to query the backend API for the occupancy status of rooms. It updates the `RoomMap.jsx` component to fetch and display the availability data on the room popups. It also adds a new localization key `rooms.map.roomAvailabilityUnknown` for both German and English languages. It removes some unnecessary console log statements and adds some commented-out code for debugging purposes.

> _To show rooms on the map with ease_
> _We added `getRoomAvailability`_
> _But we left some debug code_
> _That should not be in prod mode_
> _So please remove the console logs, pretty please_

### Walkthrough
*  Added a new feature to display the availability information of rooms on the map ([link](https://github.com/neuland-ingolstadt/neuland.app/pull/354/files?diff=unified&w=0#diff-1b6530a85236f66bb5fc35f177c7fa65fa39f0e26fff16cb66eb69c881f66065L15-R15), [link](https://github.com/neuland-ingolstadt/neuland.app/pull/354/files?diff=unified&w=0#diff-1b6530a85236f66bb5fc35f177c7fa65fa39f0e26fff16cb66eb69c881f66065R61), [link](https://github.com/neuland-ingolstadt/neuland.app/pull/354/files?diff=unified&w=0#diff-1b6530a85236f66bb5fc35f177c7fa65fa39f0e26fff16cb66eb69c881f66065R130-R131), [link](https://github.com/neuland-ingolstadt/neuland.app/pull/354/files?diff=unified&w=0#diff-1b6530a85236f66bb5fc35f177c7fa65fa39f0e26fff16cb66eb69c881f66065R196), [link](https://github.com/neuland-ingolstadt/neuland.app/pull/354/files?diff=unified&w=0#diff-1b6530a85236f66bb5fc35f177c7fa65fa39f0e26fff16cb66eb69c881f66065R224-R232), [link](https://github.com/neuland-ingolstadt/neuland.app/pull/354/files?diff=unified&w=0#diff-1b6530a85236f66bb5fc35f177c7fa65fa39f0e26fff16cb66eb69c881f66065R274-R338), [link](https://github.com/neuland-ingolstadt/neuland.app/pull/354/files?diff=unified&w=0#diff-cec180dc2d7f04377a0dd9436facdf6de328b7c0c8d969bc701be36842dca60aR194-R260), [link](https://github.com/neuland-ingolstadt/neuland.app/pull/354/files?diff=unified&w=0#diff-77b216d2e136b509f8aa884bf80652d935617d86ca210980ba498164faacab8bR17), [link](https://github.com/neuland-ingolstadt/neuland.app/pull/354/files?diff=unified&w=0#diff-7691f254ea1919a39e2b29148e081d73457787a40afcb147bd259bd741bd14a3R17))
  * Defined a new async function `getRoomAvailability` in `rooms-utils.js` that fetches the free time slots for a given list of rooms and an optional date ([link](https://github.com/neuland-ingolstadt/neuland.app/pull/354/files?diff=unified&w=0#diff-cec180dc2d7f04377a0dd9436facdf6de328b7c0c8d969bc701be36842dca60aR194-R260))
  * Imported `getRoomAvailability` in `RoomMap.jsx` and declared a new state variable `roomAvailabilityList` using the `useState` hook ([link](https://github.com/neuland-ingolstadt/neuland.app/pull/354/files?diff=unified&w=0#diff-1b6530a85236f66bb5fc35f177c7fa65fa39f0e26fff16cb66eb69c881f66065L15-R15), [link](https://github.com/neuland-ingolstadt/neuland.app/pull/354/files?diff=unified&w=0#diff-1b6530a85236f66bb5fc35f177c7fa65fa39f0e26fff16cb66eb69c881f66065R61))
  * Invoked `loadRoomAvailability` in the `useEffect` hook in `RoomMap.jsx`, passing the `filtered` variable as an argument. This function calls `getRoomAvailability` and updates the `roomAvailabilityList` state variable with the data from the backend API ([link](https://github.com/neuland-ingolstadt/neuland.app/pull/354/files?diff=unified&w=0#diff-1b6530a85236f66bb5fc35f177c7fa65fa39f0e26fff16cb66eb69c881f66065R130-R131), [link](https://github.com/neuland-ingolstadt/neuland.app/pull/354/files?diff=unified&w=0#diff-1b6530a85236f66bb5fc35f177c7fa65fa39f0e26fff16cb66eb69c881f66065R274-R338))
  * Modified the `renderRoom` function in `RoomMap.jsx` to display the availability information of the room in the popup using the `t` function and the `rooms.map.freeFromUntil` key from the `rooms.json` files. The `roomAvailabilityList` state variable is accessed using the `entry.properties.Raum` key, which is the room name. The `from` and `until` properties of the availability object are formatted as hours and minutes. If the availability object is undefined or null, the text 'Error' is displayed instead ([link](https://github.com/neuland-ingolstadt/neuland.app/pull/354/files?diff=unified&w=0#diff-1b6530a85236f66bb5fc35f177c7fa65fa39f0e26fff16cb66eb69c881f66065R224-R232))

### CODE REVIEW

**Review by BuildmodeOne — CHANGES_REQUESTED** [review]
image

Time of testing: 29.10.2023 - 22:46

File: rogue-thi-app/components/RoomMap.jsx [hunk 1]
```diff
@@ -258,6 +271,71 @@ export default function RoomMap ({ highlight, roomData }) {
     )
   }
 
+  const devmode = false //!
+  // useEffect(() => {
```
Comment (BuildmodeOne) [inline rogue-thi-app/components/RoomMap.jsx:267]: please remove the comments :)

File: rogue-thi-app/components/RoomMap.jsx [hunk 2]
```diff
@@ -217,6 +221,15 @@ export default function RoomMap ({ highlight, roomData }) {
             <>
               <br />
               {t('rooms.map.occupied')}
+              <br />
+              {/* {t('rooms.map.freeFromUntil', {
```
Comment (BuildmodeOne) [inline rogue-thi-app/components/RoomMap.jsx]: here as well

File: rogue-thi-app/components/RoomMap.jsx [hunk 3]
```diff
@@ -269,6 +347,23 @@ export default function RoomMap ({ highlight, roomData }) {
           isInvalid={filteredRooms.length === 0}
           ref={searchField}
         />
+
```
Comment (BuildmodeOne) [inline rogue-thi-app/components/RoomMap.jsx:339]: and here

File: rogue-thi-app/lib/backend-utils/rooms-utils.js [hunk 4]
```diff
@@ -189,6 +190,73 @@ export async function searchRooms (beginDate, endDate, building = BUILDINGS_ALL)
...
+ * @param {Array} [roomRequestList] Building name (e.g. `["G215"]`)
+ * @returns {object[]}
+ */
+export async function getRoomAvailability (roomRequestList, beginDate) {
+  if (beginDate === undefined) {
+    beginDate = new Date()
+    beginDate.setHours(0, 0, 0, 0)
+  }
+
+  const devmode = false //!
```
Comment (BuildmodeOne) [inline rogue-thi-app/lib/backend-utils/rooms-utils.js:205]: Good for testing purposes but shouldn't be in the project.

File: rogue-thi-app/lib/backend-utils/rooms-utils.js [hunk 5]
```diff
@@ -189,6 +190,73 @@ export async function searchRooms (beginDate, endDate, building = BUILDINGS_ALL)
...
+              if (String(thisRoom2).toLowerCase() === thisRoom.toLocaleLowerCase()) {
+                raumFrei.push({ von, bis })
+              }
+            }
+          }
+        }
+      }
+    }
+
+    // Zeiten verbinden //? nochmal verbinden? Kann ich nicht testen, da es immer so pausen dazwischen gibt
```
Comment (BuildmodeOne) [inline rogue-thi-app/lib/backend-utils/rooms-utils.js:242]: Please keep the comments in English.

File: rogue-thi-app/lib/backend-utils/rooms-utils.js [hunk 6]
```diff
@@ -189,6 +190,73 @@ export async function searchRooms (beginDate, endDate, building = BUILDINGS_ALL)
...
+        raumFreiCombined.splice(index - offset, 1)
+        raumFreiCombined.splice(index - offset, 1, { von: raumFrei[index]['von'], bis: raumFrei[index + 1]['bis'] })
+        offset++
+      }
+    }
+    // console.log(raumFreiCombined)
+    roomList[thisRoom] = raumFreiCombined
+  }
+
+  // console.log(roomList)
```
Comment (BuildmodeOne) [inline rogue-thi-app/lib/backend-utils/rooms-utils.js:256]: please remove all debug `console.log` calls :)

File: rogue-thi-app/components/RoomMap.jsx [hunk 7]
```diff
@@ -258,6 +271,71 @@ export default function RoomMap ({ highlight, roomData }) {
...
+    const roomAvailabilityData = await getRoomAvailability(roomRequestList)
+    // console.log(roomAvailabilityData)
+
+    for (const room in roomAvailabilityData) {
+      let thisRoomAvailability = {
+        from: t('rooms.map.roomAvailabilityUnknown'),
+        until: t('rooms.map.roomAvailabilityUnknown')
+      }
+      if (roomAvailabilityData[room]?.length) {
+        let fromDate
```
Comment (BuildmodeOne) [inline rogue-thi-app/components/RoomMap.jsx:282]: Not biggest fan of `let` but this is just my taste.

**Review by BuildmodeOne — CHANGES_REQUESTED** [review]

**Comment by pl0ss** [comment]
Bildschirmfoto 2023-10-30 um 20 38 18

This is how it would look on October 30th at 12 o'clock


Bildschirmfoto 2023-10-30 um 20 44 43

and at 20 o'clock

File: rogue-thi-app/components/RoomMap.jsx [hunk 8]
```diff
@@ -352,7 +392,7 @@ export default function RoomMap ({ highlight, roomData }) {
               <LayersControl.BaseLayer
                 key={floorName + (searchText || 'empty-search')}
                 name={translateFloors(floorName)}
-                checked={i === filteredFloorOrder.length - 2}
+                checked={i === filteredFloorOrder.length - (searchText ? 1 : 2)}
```
Comment (Robert27) [inline rogue-thi-app/components/RoomMap.jsx:395]: When searching for `I` this shows the basement by default. 
However basements are not that relevant, which is why we should still display the ground floor.

**Review by Robert27 — APPROVED** [review]

**Review by BuildmodeOne — APPROVED** [review]

**Review by Robert27 — APPROVED** [review]

**Pull request merged**